In [74]:
import pandas as pd
import numpy as np

In [75]:
## Reading csv
data = pd.read_csv('MScFE560_FM_GWP1_Data.csv')
avg_returns = data[:1]
std_dev = data[1:2]
cov_matrix = data[6:]
std_dev

Unnamed: 0         AAPL         AMZN        NFLX         META        GOOG  \
1     St Dev  0.009320453  0.009556763  0.01340882  0.011581856  0.00837577   

   Unnamed: 6                           Unnamed: 7  
1         NaN  Standard deviation of daily returns

In [76]:
avg_returns

Unnamed: 0         AAPL         AMZN          NFLX          META  \
0       Mean  0.044857944  0.010361979  -0.029674829  -0.006934285   

          GOOG  Unnamed: 6            Unnamed: 7  
0  0.029035134         NaN  Average daily return

In [77]:
cov_matrix['rows'] = ['AAPL','AMZN','NFLX','META','GOOG']
cov_matrix

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Unnamed: 0         AAPL         AMZN         NFLX         META  \
6        AAPL            1  0.660738586  0.460041141  0.594784944   
7        AMZN  0.660738586            1  0.593812086  0.626197627   
8        NFLX  0.460041141  0.593812086            1  0.515892088   
9        META  0.594784944  0.626197627  0.515892088            1   
10      GOOGL  0.698431255  0.679322691  0.492697083  0.668023785   

           GOOG  Unnamed: 6                           Unnamed: 7  rows  
6   0.698431255         NaN  Correlation Matrix of daily returns  AAPL  
7   0.679322691         NaN                                  NaN  AMZN  
8   0.492697083         NaN                                  NaN  NFLX  
9   0.668023785         NaN                                  NaN  META  
10            1         NaN                                  NaN  GOOG

## Analysis for Portfolio A

In [78]:
stocks = ['AAPL','AMZN','NFLX','META','GOOG']

In [79]:
for i in stocks:
    for j in stocks:
        if(i==j):
            continue
        best_coeff_var = 10**8
        best_wa = 0
        best_wb = 0
        for wa in np.arange(-2,1,0.01):
            wb = 1 - wa
            if(max(wa,wb)<=0):
                continue
            expected_return = wa * float(avg_returns[i]) + wb * float(avg_returns[j])
            if(expected_return<=0):
                continue
            prt_std_dev = (wa**2 * float(std_dev[i])**2 + wb**2 * float(std_dev[j])**2 + 2*wa*wb*float(std_dev[i])*float(std_dev[j])*float(cov_matrix[cov_matrix.rows==i][j]))**0.5
            coeff_variation = prt_std_dev / expected_return
            if(coeff_variation < best_coeff_var):
                best_coeff_var = coeff_variation
                best_wa = wa
                best_wb = wb
        print(f"With wa = {best_wa} and wb = {best_wb} weights has the best coeff_var for {i} and {j} and is {best_coeff_var}")
    

With wa = 0.9900000000000029 and wb = 0.009999999999997122 weights has the best coeff_var for AAPL and AMZN and is 0.20871821485534628
With wa = 0.9900000000000029 and wb = 0.009999999999997122 weights has the best coeff_var for AAPL and NFLX and is 0.21059049063253216
With wa = 0.9900000000000029 and wb = 0.009999999999997122 weights has the best coeff_var for AAPL and META and is 0.20966615436122715
With wa = 0.9500000000000028 and wb = 0.04999999999999716 weights has the best coeff_var for AAPL and GOOG and is 0.2076809231375243
With wa = -0.9999999999999991 and wb = 1.9999999999999991 weights has the best coeff_var for AMZN and AAPL and is 0.17972368022248406
With wa = 0.9900000000000029 and wb = 0.009999999999997122 weights has the best coeff_var for AMZN and NFLX and is 0.9578198366256915
With wa = 0.9900000000000029 and wb = 0.009999999999997122 weights has the best coeff_var for AMZN and META and is 0.9357280400703928
With wa = -0.6899999999999988 and wb = 1.6899999999999988 we

#### With wa = -1.1199999999999992 and wb = 2.119999999999999 weights has the best coeff_var for NFLX and AAPL and is 0.1442996889459886

## For portfolio B


In [81]:
for i in stocks:
    for j in stocks:
        if(i==j):
            continue
        best_coeff_var = 10**8
        best_wa = 0
        best_wb = 0
        for wa in np.arange(0.01,1,0.01):
            wb = 1 - wa
            if(max(wa,wb)<=0):
                continue
            expected_return = wa * float(avg_returns[i]) + wb * float(avg_returns[j])
            if(expected_return<=0):
                continue
            prt_std_dev = (wa**2 * float(std_dev[i])**2 + wb**2 * float(std_dev[j])**2 + 2*wa*wb*float(std_dev[i])*float(std_dev[j])*float(cov_matrix[cov_matrix.rows==i][j]))**0.5
            coeff_variation = prt_std_dev / expected_return
            if(coeff_variation < best_coeff_var):
                best_coeff_var = coeff_variation
                best_wa = wa
                best_wb = wb
        print(f"With wa = {best_wa} and wb = {best_wb} weights has the best coeff_var for {i} and {j} and is {best_coeff_var}")
    

With wa = 0.99 and wb = 0.010000000000000009 weights has the best coeff_var for AAPL and AMZN and is 0.20871821485534653
With wa = 0.99 and wb = 0.010000000000000009 weights has the best coeff_var for AAPL and NFLX and is 0.21059049063253302
With wa = 0.99 and wb = 0.010000000000000009 weights has the best coeff_var for AAPL and META and is 0.2096661543612277
With wa = 0.9500000000000001 and wb = 0.04999999999999993 weights has the best coeff_var for AAPL and GOOG and is 0.20768092313752426
With wa = 0.01 and wb = 0.99 weights has the best coeff_var for AMZN and AAPL and is 0.20871821485534653
With wa = 0.99 and wb = 0.010000000000000009 weights has the best coeff_var for AMZN and NFLX and is 0.9578198366257024
With wa = 0.99 and wb = 0.010000000000000009 weights has the best coeff_var for AMZN and META and is 0.9357280400703966
With wa = 0.01 and wb = 0.99 weights has the best coeff_var for AMZN and GOOG and is 0.28969464911004084
With wa = 0.01 and wb = 0.99 weights has the best coef

## With wa = 0.05 and wb = 0.95 weights has the best coeff_var for GOOG and AAPL and is 0.2076809231375243

# Portfolio C

In [ ]:
for i in stocks:
    for j in stocks:
        for k in stocks:
            if(i==j or j==k or i==k):
                continue
            best_coeff_var = 10**8
            best_wa = 0
            best_wb = 0
            best_wc = 0
            for wa in np.arange(-2,1,0.01):
                for wb in np.arange(-2,1,0.01):    
                    wc = 1 - wa - wb
                    if(max(wa,wb,wc)<=0):
                        continue
                    expected_return = wa * float(avg_returns[i]) + \
                                      wb * float(avg_returns[j]) + \
                                      wc * float(avg_returns[k])

                    if(expected_return<=0):
                        continue
                    prt_std_dev = (wa**2 * float(std_dev[i])**2 + 
                                   wb**2 * float(std_dev[j])**2 +
                                   wc**2 * float(std_dev[k])**2 + 
                                   2*wa*wb*float(std_dev[i])*float(std_dev[j])*float(cov_matrix[cov_matrix.rows==i][j]) + 
                                   2*wa*wc*float(std_dev[i])*float(std_dev[k])*float(cov_matrix[cov_matrix.rows==i][k]) +
                                   2*wc*wb*float(std_dev[k])*float(std_dev[j])*float(cov_matrix[cov_matrix.rows==k][j])
                                   )**0.5
                    coeff_variation = prt_std_dev / expected_return
                    if(coeff_variation < best_coeff_var):
                        best_coeff_var = coeff_variation
                        best_wa = wa
                        best_wb = wb
                        best_wc = wc
            print(f"With wa = {best_wa}, wb = {best_wb} and wc = {best_wc} weights has the best coeff_var for {i},{j} and {k}, and is {best_coeff_var}")
    

With wa = 0.9900000000000029, wb = 0.6600000000000024 and wc = -0.6500000000000052 weights has the best coeff_var for AAPL,AMZN and NFLX, and is 0.16685351890144623
With wa = 0.9900000000000029, wb = 0.44000000000000217 and wc = -0.43000000000000504 weights has the best coeff_var for AAPL,AMZN and META, and is 0.19056451082844725
With wa = 0.9900000000000029, wb = -0.7399999999999989 and wc = 0.749999999999996 weights has the best coeff_var for AAPL,AMZN and GOOG, and is 0.1784068358930561
With wa = 0.9900000000000029, wb = -0.6499999999999988 and wc = 0.6599999999999959 weights has the best coeff_var for AAPL,NFLX and AMZN, and is 0.1668535189014462
With wa = 0.9900000000000029, wb = -0.24999999999999845 and wc = 0.25999999999999557 weights has the best coeff_var for AAPL,NFLX and META, and is 0.199524762574703
With wa = 0.9900000000000029, wb = -0.8099999999999989 and wc = 0.8199999999999961 weights has the best coeff_var for AAPL,NFLX and GOOG, and is 0.14260782209975553
With wa = 0

## With wa = 0.9900000000000029, wb = 0.8200000000000025 and wc = -0.8100000000000054 weights has the best coeff_var for AAPL,GOOG and NFLX, and is 0.14260782209975553